futures-util/src/async_stream/mod.rs

```diff
@@ -0,0 +1,83 @@
+//! Await
+//!
+//! This module contains a number of functions and combinators for working
+//! with `async`/`await` code.
+
+#[macro_use]
+mod await_item;
+pub use self::await_item::*;
+
+#[macro_use]
+mod stream_yield;
+pub use self::stream_yield::*;
+
+use futures_core::future::Future;
+use futures_core::stream::Stream;
+use std::future::{self, get_task_context, set_task_context};
+use std::marker::{PhantomData, PhantomPinned};
+use std::ops::{Generator, GeneratorState};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// Wrap a generator in a stream.
+///
+/// This function returns a `GenStream` underneath, but hides it in `impl Trait` to give
+/// better error messages (`impl Stream` rather than `GenStream<[closure.....]>`).
+pub fn from_generator<U, T>(x: T) -> impl Stream<Item = U>
+where
+    T: Generator<Yield = Poll<U>, Return = ()>,
+{
+    GenStream { gen: x, done: false, _phantom: PhantomData, _pin: PhantomPinned }
+}
+
+/// A wrapper around generators used to implement `Stream` for `async`/`await` code.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+struct GenStream<U, T: Generator<Yield = Poll<U>, Return = ()>> {
+    gen: T,
+    // If resuming after Complete, std generators panic. This is natural when using the generators,
+    // but in the streams, we may want to call `poll_next` after `poll_next` returns `None`
+    // because it is possible to call `next` after `next` returns `None` in many iterators.
+    done: bool,
+    _phantom: PhantomData<U>,
+    // We rely on the fact that async/await streams are immovable in order to create
+    // self-referential borrows in the underlying generator.
+    _pin: PhantomPinned,
```

Comment:
You shouldn't need this, `T: ?Unpin` so `Self: ?Unpin` via the structural rules (although because of `PhantomData<U>` you might be getting the `Unpin`ness of `U` as well), if the generator transform ever allows non-self-referential `static`-marked generators to implement `Unpin` automatically then this could potentially generate `impl Stream + Unpin`.